[Feat] #590 Level 2: Echo Server Sample

## Parent Epic
Part of #590
Depends on #591, #592

## Summary

Create a DICOM Echo Server (Verification SCP) sample demonstrating network communication fundamentals: server configuration, association handling, and C-ECHO service.

## Background

### Current State
- `examples/echo_scp` exists but lacks educational comments
- No progressive learning from Level 1

### Target State
- Well-documented Echo server
- Explains association state machine
- Demonstrates event callbacks
- Testable with DCMTK echoscu

## Learning Objectives

After completing this sample, developers will understand:

1. **DICOM Server Configuration** - AE Title, port, timeouts, PDU size
2. **Association Management** - States, negotiation, presentation contexts
3. **Service Class Provider (SCP)** - Handling incoming DIMSE messages
4. **C-ECHO Operation** - Verification service for connectivity testing

## Technical Specification

### Sample Code Structure

```cpp
// samples/02_echo_server/main.cpp
#include "pacs/network/dicom_server.hpp"
#include "pacs/network/server_config.hpp"
#include "pacs/network/association.hpp"
#include "pacs/services/verification_scp.hpp"
#include "signal_handler.hpp"
#include "console_utils.hpp"

#include <iostream>
#include <memory>

namespace net = pacs::network;
namespace svc = pacs::services;

int main(int argc, char* argv[]) {
    samples::print_header("Echo Server - Level 2 Sample");
    
    // ═══════════════════════════════════════════════════════════
    // Part 1: Server Configuration
    // ═══════════════════════════════════════════════════════════
    // The server needs basic DICOM network parameters:
    // - AE Title: Application Entity identifier (max 16 chars)
    // - Port: TCP port to listen on (11112 is common)
    // - Max PDU Length: Maximum Protocol Data Unit size
    // - Timeouts: Connection and idle timeouts
    
    std::cout << "\n=== Part 1: Server Configuration ===\n\n";
    
    net::server_config config;
    config.ae_title = "ECHO_SCP";                          // Our AE Title
    config.port = argc > 1 ? std::stoi(argv[1]) : 11112;   // Listening port
    config.max_associations = 10;                           // Concurrent connections
    config.idle_timeout = std::chrono::seconds(60);        // Idle disconnect
    config.max_pdu_length = 16384;                         // 16KB PDU size
    
    // Implementation identification (optional but recommended)
    config.implementation_class_uid = "1.2.410.200001.1.1";
    config.implementation_version_name = "PACS_SAMPLE_1.0";
    
    std::cout << "Configuration:\n";
    std::cout << "  AE Title:         " << config.ae_title << "\n";
    std::cout << "  Port:             " << config.port << "\n";
    std::cout << "  Max Associations: " << config.max_associations << "\n";
    std::cout << "  Max PDU Length:   " << config.max_pdu_length << " bytes\n";
    
    // ═══════════════════════════════════════════════════════════
    // Part 2: Create Server and Register Services
    // ═══════════════════════════════════════════════════════════
    // DICOM servers host one or more Service Class Providers (SCPs).
    // Each SCP handles specific DIMSE operations.
    // Verification SCP handles C-ECHO requests.
    
    std::cout << "\n=== Part 2: Server & Service Setup ===\n\n";
    
    // Create server instance
    auto server = std::make_unique<net::dicom_server>(config);
    
    // Create and register Verification SCP (C-ECHO handler)
    // The Verification SOP Class UID is: 1.2.840.10008.1.1
    auto verification_service = std::make_shared<svc::verification_scp>();
    server->register_service(verification_service);
    
    std::cout << "Registered services:\n";
    std::cout << "  ✓ Verification SCP (C-ECHO)\n";
    
    // ═══════════════════════════════════════════════════════════
    // Part 3: Event Callbacks
    // ═══════════════════════════════════════════════════════════
    // The server emits events for association lifecycle:
    // - Association established (after A-ASSOCIATE-AC sent)
    // - Association released (after A-RELEASE)
    // - Association aborted (on A-ABORT or error)
    // - Errors (network/protocol errors)
    
    std::cout << "\n=== Part 3: Event Handlers ===\n\n";
    
    // Track connection statistics
    std::atomic<int> connection_count{0};
    std::atomic<int> echo_count{0};
    
    server->on_association_established(
        [&connection_count](const net::association& assoc) {
            connection_count++;
            std::cout << "[CONNECT] " << assoc.calling_ae() 
                      << " -> " << assoc.called_ae() << "\n";
            std::cout << "          State: " << to_string(assoc.state()) << "\n";
        }
    );
    
    server->on_association_released(
        [&connection_count](const net::association& assoc) {
            connection_count--;
            std::cout << "[RELEASE] " << assoc.calling_ae() << " disconnected\n";
        }
    );
    
    server->on_error(
        [](const std::string& error_msg) {
            std::cerr << "[ERROR] " << error_msg << "\n";
        }
    );
    
    // Hook into C-ECHO handling for statistics
    verification_service->set_echo_callback(
        [&echo_count](const std::string& calling_ae) {
            echo_count++;
            std::cout << "[C-ECHO] Request from " << calling_ae 
                      << " (total: " << echo_count << ")\n";
        }
    );
    
    std::cout << "Event handlers registered.\n";
    
    // ═══════════════════════════════════════════════════════════
    // Part 4: Start Server
    // ═══════════════════════════════════════════════════════════
    
    std::cout << "\n=== Part 4: Running Server ===\n\n";
    
    // Install signal handler for graceful shutdown
    samples::scoped_signal_handler sig_handler([&server]() {
        std::cout << "\nShutdown requested...\n";
        server->stop();
    });
    
    // Start the server (non-blocking)
    auto start_result = server->start();
    if (start_result.is_err()) {
        std::cerr << "Failed to start server: " 
                  << start_result.error().message << "\n";
        return 1;
    }
    
    std::cout << "╔════════════════════════════════════════════╗\n";
    std::cout << "║        Echo Server Running                 ║\n";
    std::cout << "╠════════════════════════════════════════════╣\n";
    std::cout << "║  Test with:                                ║\n";
    std::cout << "║  echoscu -aec ECHO_SCP localhost " 
              << config.port << "     ║\n";
    std::cout << "║                                            ║\n";
    std::cout << "║  Press Ctrl+C to stop                      ║\n";
    std::cout << "╚════════════════════════════════════════════╝\n\n";
    
    // Wait for shutdown signal
    sig_handler.wait();
    
    // Print final statistics
    std::cout << "\n=== Final Statistics ===\n";
    std::cout << "  Total C-ECHO requests: " << echo_count << "\n";
    std::cout << "  Server stopped gracefully.\n";
    
    return 0;
}
```

## Implementation Tasks

- [ ] Create `samples/02_echo_server/` directory
- [ ] Implement `main.cpp` with all 4 parts
- [ ] Add optional echo callback to verification_scp (if not exists)
- [ ] Create `CMakeLists.txt`
- [ ] Create `README.md` with test instructions
- [ ] Test with DCMTK echoscu

## File Structure

```
samples/02_echo_server/
├── CMakeLists.txt
├── main.cpp
└── README.md
```

## Test Commands

```bash
# Build
cmake -B build -DPACS_BUILD_SAMPLES=ON
cmake --build build --target sample_02_echo_server

# Run server
./build/samples/echo_server 11112

# Test (in another terminal)
echoscu -v -aec ECHO_SCP localhost 11112

# Multiple tests
for i in {1..10}; do echoscu -aec ECHO_SCP localhost 11112; done
```

## Acceptance Criteria

- [ ] Server starts and listens on configured port
- [ ] echoscu connects and receives success response
- [ ] Event callbacks fire correctly
- [ ] Graceful shutdown on Ctrl+C
- [ ] Code well-commented for learning
- [ ] README with test instructions

## Dependencies

- #591 (Common utilities)
- #592 (Level 1 concepts)
- pacs::network
- pacs::services

## Estimated Size

~200-250 LOC (S size)

--- a/samples/02_echo_server/CMakeLists.txt
+++ b/samples/02_echo_server/CMakeLists.txt
@@ -1,0 +1,24 @@
+# samples/02_echo_server/CMakeLists.txt
+# Level 2: Echo Server - DICOM network communication fundamentals
+
+add_executable(sample_02_echo_server
+    main.cpp
+)
+
+target_link_libraries(sample_02_echo_server PRIVATE
+    pacs_network
+    pacs_services
+    pacs_samples_common
+)
+
+# Require C++20
+target_compile_features(sample_02_echo_server PRIVATE cxx_std_20)
+
+# Set output properties
+set_target_properties(sample_02_echo_server PROPERTIES
+    OUTPUT_NAME "echo_server"
+    RUNTIME_OUTPUT_DIRECTORY "${CMAKE_BINARY_DIR}/samples"
+)
+
+# Add alias for consistent naming
+add_executable(pacs::samples::echo_server ALIAS sample_02_echo_server)

--- a/samples/02_echo_server/README.md
+++ b/samples/02_echo_server/README.md
@@ -1,0 +1,173 @@
+# Level 2: Echo Server
+
+A DICOM Echo Server (Verification SCP) sample demonstrating network communication fundamentals. This builds on Level 1 concepts and introduces DICOM networking.
+
+## Learning Objectives
+
+By completing this sample, you will understand:
+
+- **Server Configuration** - AE Title, port, timeouts, PDU size
+- **Association Management** - Connection states, negotiation flow
+- **Service Class Provider (SCP)** - Handling incoming DIMSE requests
+- **C-ECHO Operation** - Verification service for connectivity testing
+- **Graceful Shutdown** - Signal handling and resource cleanup
+
+## Prerequisites
+
+- Completed Level 1: Hello DICOM
+- C++20 compatible compiler (GCC 11+, Clang 14+, MSVC 2022+)
+- pacs_system library built with samples enabled
+- DCMTK installed (for testing with `echoscu`)
+
+## Build & Run
+
+```bash
+# Configure with samples enabled
+cmake -B build -DPACS_BUILD_SAMPLES=ON
+
+# Build this sample
+cmake --build build --target sample_02_echo_server
+
+# Run the server (default port 11112)
+./build/samples/echo_server
+
+# Or specify a custom port
+./build/samples/echo_server 11113
+```
+
+## Testing
+
+With the server running, open another terminal and test with DCMTK:
+
+```bash
+# Single C-ECHO test
+echoscu -v -aec ECHO_SCP localhost 11112
+
+# Multiple tests (stress test)
+for i in {1..10}; do echoscu -aec ECHO_SCP localhost 11112; done
+
+# Verbose output showing DIMSE details
+echoscu -d -aec ECHO_SCP localhost 11112
+```
+
+### Expected Output
+
+**Server side:**
+```
+[14:30:22.456] [CONNECT] ECHOSCU -> ECHO_SCP (active: 1)
+[14:30:22.458] [RELEASE] ECHOSCU disconnected (active: 0)
+```
+
+**Client side (echoscu -v):**
+```
+I: Requesting Association
+I: Association Accepted
+I: Sending Echo Request
+I: Received Echo Response (Status: Success)
+I: Releasing Association
+```
+
+## Key Concepts Explained
+
+### DICOM Association
+
+An association is a logical connection between two DICOM applications:
+
+```
+   SCU (Client)                          SCP (Server)
+       │                                     │
+       │  A-ASSOCIATE-RQ ─────────────────►  │  Request association
+       │                                     │
+       │  ◄───────────────── A-ASSOCIATE-AC  │  Accept association
+       │                                     │
+       │  C-ECHO-RQ ──────────────────────►  │  Ping request
+       │                                     │
+       │  ◄────────────────────── C-ECHO-RSP │  Ping response
+       │                                     │
+       │  A-RELEASE-RQ ───────────────────►  │  Request release
+       │                                     │
+       │  ◄─────────────────── A-RELEASE-RP  │  Confirm release
+       │                                     │
+```
+
+### Server Configuration
+
+Key parameters for DICOM servers:
+
+| Parameter | Description | Typical Value |
+|-----------|-------------|---------------|
+| AE Title | Application Entity identifier | 16 chars max |
+| Port | TCP listening port | 104 (standard), 11112 (alternate) |
+| Max PDU Size | Maximum data unit size | 16384 - 65536 bytes |
+| Idle Timeout | Disconnect after inactivity | 30 - 300 seconds |
+| Max Associations | Concurrent connection limit | 10 - 100 |
+
+### Service Registration
+
+```cpp
+// Create server with configuration
+net::dicom_server server{config};
+
+// Register Verification SCP (handles C-ECHO)
+server.register_service(std::make_shared<svc::verification_scp>());
+
+// You can register multiple services:
+// server.register_service(std::make_shared<svc::storage_scp>(path));
+// server.register_service(std::make_shared<svc::query_scp>(db));
+```
+
+### Event Callbacks
+
+```cpp
+// Monitor association lifecycle
+server.on_association_established([](const net::association& assoc) {
+    std::cout << "Connected: " << assoc.calling_ae() << "\n";
+});
+
+server.on_association_released([](const net::association& assoc) {
+    std::cout << "Disconnected: " << assoc.calling_ae() << "\n";
+});
+
+server.on_error([](const std::string& error) {
+    std::cerr << "Error: " << error << "\n";
+});
+```
+
+## Troubleshooting
+
+### Port Already in Use
+
+```bash
+# Find process using the port
+lsof -i :11112
+
+# Or use a different port
+./build/samples/echo_server 11113
+```
+
+### Connection Refused
+
+- Ensure the server is running
+- Check firewall settings
+- Verify the port number matches
+
+### AE Title Mismatch
+
+DCMTK's `echoscu` uses `-aec` to specify the called (server) AE Title:
+```bash
+echoscu -aec ECHO_SCP localhost 11112
+```
+
+## Next Steps
+
+Proceed to **Level 3: Storage Server** to learn:
+- C-STORE operation for receiving DICOM images
+- File storage and organization
+- Transfer syntax handling
+- IOD (Information Object Definition) validation
+
+## Related Documentation
+
+- [DICOM PS3.7 - Message Exchange](http://dicom.nema.org/medical/dicom/current/output/chtml/part07/PS3.7.html)
+- [DICOM PS3.8 - Network Communication](http://dicom.nema.org/medical/dicom/current/output/chtml/part08/PS3.8.html)
+- [DICOM PS3.4 Section A.4 - Verification Service](http://dicom.nema.org/medical/dicom/current/output/chtml/part04/sect_A.4.html)

--- a/samples/02_echo_server/main.cpp
+++ b/samples/02_echo_server/main.cpp
@@ -1,0 +1,302 @@
+/**
+ * @file main.cpp
+ * @brief Level 2 Sample: Echo Server - DICOM network communication fundamentals
+ *
+ * This sample demonstrates DICOM networking concepts:
+ * - Server configuration (AE Title, port, timeouts)
+ * - Association management (connection, negotiation, release)
+ * - Service Class Provider (SCP) pattern
+ * - C-ECHO operation (Verification service)
+ *
+ * After completing this sample, you will understand how to:
+ * 1. Configure a DICOM server with proper network parameters
+ * 2. Register SCP services to handle DIMSE requests
+ * 3. Set up callbacks for association lifecycle events
+ * 4. Implement graceful shutdown handling
+ *
+ * @see DICOM PS3.7 Section 9.1 - C-ECHO Service
+ * @see DICOM PS3.8 - Network Communication Support
+ */
+
+#include "console_utils.hpp"
+#include "signal_handler.hpp"
+
+#include <pacs/network/dicom_server.hpp>
+#include <pacs/network/server_config.hpp>
+#include <pacs/network/association.hpp>
+#include <pacs/services/verification_scp.hpp>
+
+#include <atomic>
+#include <chrono>
+#include <cstdlib>
+#include <iomanip>
+#include <iostream>
+#include <memory>
+#include <string>
+
+namespace net = pacs::network;
+namespace svc = pacs::services;
+
+namespace {
+
+/**
+ * @brief Format timestamp for logging
+ * @return Current time as formatted string
+ */
+std::string current_timestamp() {
+    auto now = std::chrono::system_clock::now();
+    auto time_t_now = std::chrono::system_clock::to_time_t(now);
+    auto ms = std::chrono::duration_cast<std::chrono::milliseconds>(
+        now.time_since_epoch()) % 1000;
+
+    std::ostringstream oss;
+    oss << std::put_time(std::localtime(&time_t_now), "%H:%M:%S");
+    oss << '.' << std::setfill('0') << std::setw(3) << ms.count();
+    return oss.str();
+}
+
+}  // namespace
+
+int main(int argc, char* argv[]) {
+    pacs::samples::print_header("Echo Server - Level 2 Sample");
+
+    // ===========================================================================
+    // Part 1: Server Configuration
+    // ===========================================================================
+    // DICOM servers require several network parameters:
+    // - AE Title: Application Entity identifier (max 16 characters)
+    //   This is like a hostname in DICOM networking
+    // - Port: TCP port to listen on (11112 is a common non-privileged port)
+    // - Max PDU Length: Maximum Protocol Data Unit size for data transfer
+    // - Timeouts: Connection and idle timeouts for resource management
+    // - Max Associations: Limit concurrent connections to prevent overload
+
+    pacs::samples::print_section("Part 1: Server Configuration");
+
+    std::cout << "DICOM servers use these key parameters:\n";
+    std::cout << "  - AE Title:    Application Entity identifier (like hostname)\n";
+    std::cout << "  - Port:        TCP listening port\n";
+    std::cout << "  - Max PDU:     Maximum data unit size per transfer\n";
+    std::cout << "  - Timeouts:    Connection and idle timeouts\n\n";
+
+    // Parse optional port argument
+    uint16_t port = 11112;  // Default port
+    if (argc > 1) {
+        try {
+            int port_arg = std::stoi(argv[1]);
+            if (port_arg > 0 && port_arg < 65536) {
+                port = static_cast<uint16_t>(port_arg);
+            }
+        } catch (const std::exception&) {
+            // Keep default port if parsing fails
+        }
+    }
+
+    // Create server configuration
+    net::server_config config;
+    config.ae_title = "ECHO_SCP";                          // Our AE Title
+    config.port = port;                                     // Listening port
+    config.max_associations = 10;                           // Concurrent connections
+    config.idle_timeout = std::chrono::seconds(60);        // Idle disconnect
+    config.max_pdu_size = 16384;                           // 16KB PDU size
+
+    // Implementation identification (optional but recommended for interoperability)
+    // The Implementation Class UID uniquely identifies your software implementation
+    config.implementation_class_uid = "1.2.410.200001.1.1";
+    config.implementation_version_name = "PACS_SAMPLE_2.0";
+
+    pacs::samples::print_table("Server Configuration", {
+        {"AE Title", config.ae_title},
+        {"Port", std::to_string(config.port)},
+        {"Max Associations", std::to_string(config.max_associations)},
+        {"Idle Timeout", std::to_string(config.idle_timeout.count()) + " seconds"},
+        {"Max PDU Size", std::to_string(config.max_pdu_size) + " bytes"}
+    });
+
+    pacs::samples::print_success("Part 1 complete - Configuration ready!");
+
+    // ===========================================================================
+    // Part 2: Create Server and Register Services
+    // ===========================================================================
+    // DICOM servers host one or more Service Class Providers (SCPs).
+    // Each SCP handles specific DIMSE (DICOM Message Service Element) operations:
+    // - Verification SCP: Handles C-ECHO (connectivity test)
+    // - Storage SCP: Handles C-STORE (image storage)
+    // - Query SCP: Handles C-FIND (data query)
+    // - Retrieve SCP: Handles C-GET/C-MOVE (data retrieval)
+    //
+    // For this sample, we only register the Verification SCP to handle C-ECHO.
+    // C-ECHO is the simplest DICOM service - like a "ping" for DICOM.
+
+    pacs::samples::print_section("Part 2: Server & Service Setup");
+
+    std::cout << "DICOM Service Class Provider (SCP) types:\n";
+    std::cout << "  - Verification SCP: C-ECHO (connectivity test)\n";
+    std::cout << "  - Storage SCP:      C-STORE (receive images)\n";
+    std::cout << "  - Query SCP:        C-FIND (search data)\n";
+    std::cout << "  - Retrieve SCP:     C-GET/C-MOVE (retrieve data)\n\n";
+
+    // Create server instance
+    auto server = std::make_unique<net::dicom_server>(config);
+
+    // Register Verification SCP (C-ECHO handler)
+    // The Verification SOP Class UID is: 1.2.840.10008.1.1
+    // This is the standard UID for the Verification service
+    auto verification_service = std::make_shared<svc::verification_scp>();
+    server->register_service(verification_service);
+
+    std::cout << "Registered services:\n";
+    for (const auto& sop_class : server->supported_sop_classes()) {
+        if (sop_class == "1.2.840.10008.1.1") {
+            std::cout << "  - Verification SCP (1.2.840.10008.1.1)\n";
+        } else {
+            std::cout << "  - " << sop_class << "\n";
+        }
+    }
+
+    pacs::samples::print_success("Part 2 complete - Verification SCP registered!");
+
+    // ===========================================================================
+    // Part 3: Event Callbacks
+    // ===========================================================================
+    // DICOM associations go through a lifecycle:
+    // 1. Association Request (A-ASSOCIATE-RQ): Client requests connection
+    // 2. Association Accept (A-ASSOCIATE-AC): Server accepts connection
+    // 3. DIMSE Exchange: Client sends C-ECHO-RQ, server responds C-ECHO-RSP
+    // 4. Association Release (A-RELEASE): Clean disconnection
+    //
+    // The server provides callbacks for monitoring these events.
+    // This is useful for logging, statistics, and debugging.
+
+    pacs::samples::print_section("Part 3: Event Handlers");
+
+    std::cout << "DICOM Association lifecycle:\n";
+    std::cout << "  1. A-ASSOCIATE-RQ  -> Client requests connection\n";
+    std::cout << "  2. A-ASSOCIATE-AC  -> Server accepts\n";
+    std::cout << "  3. DIMSE exchange  -> C-ECHO, C-STORE, etc.\n";
+    std::cout << "  4. A-RELEASE       -> Clean disconnection\n\n";
+
+    // Track connection statistics
+    std::atomic<int> connection_count{0};
+    std::atomic<int> total_connections{0};
+
+    // Set up association established callback
+    // Called when a new association is successfully negotiated
+    server->on_association_established(
+        [&connection_count, &total_connections](const net::association& assoc) {
+            connection_count++;
+            total_connections++;
+            std::cout << "[" << current_timestamp() << "] "
+                      << pacs::samples::colors::green << "[CONNECT]"
+                      << pacs::samples::colors::reset << " "
+                      << assoc.calling_ae() << " -> " << assoc.called_ae()
+                      << " (active: " << connection_count << ")\n";
+        }
+    );
+
+    // Set up association released callback
+    // Called when an association is gracefully released
+    server->on_association_released(
+        [&connection_count](const net::association& assoc) {
+            connection_count--;
+            std::cout << "[" << current_timestamp() << "] "
+                      << pacs::samples::colors::cyan << "[RELEASE]"
+                      << pacs::samples::colors::reset << " "
+                      << assoc.calling_ae() << " disconnected"
+                      << " (active: " << connection_count << ")\n";
+        }
+    );
+
+    // Set up error callback
+    // Called when network or protocol errors occur
+    server->on_error(
+        [](const std::string& error_msg) {
+            std::cerr << "[" << current_timestamp() << "] "
+                      << pacs::samples::colors::red << "[ERROR]"
+                      << pacs::samples::colors::reset << " "
+                      << error_msg << "\n";
+        }
+    );
+
+    pacs::samples::print_success("Part 3 complete - Event handlers registered!");
+
+    // ===========================================================================
+    // Part 4: Start Server
+    // ===========================================================================
+    // The server starts listening on the configured port.
+    // It runs in the background, handling associations in a thread pool.
+    // The main thread waits for a shutdown signal (Ctrl+C).
+
+    pacs::samples::print_section("Part 4: Running Server");
+
+    // Install signal handler for graceful shutdown (Ctrl+C)
+    // scoped_signal_handler automatically cleans up when destroyed
+    pacs::samples::scoped_signal_handler sig_handler([&server]() {
+        std::cout << "\n" << pacs::samples::colors::yellow
+                  << "Shutdown signal received..."
+                  << pacs::samples::colors::reset << "\n";
+        server->stop();
+    });
+
+    // Start the server (non-blocking)
+    auto start_result = server->start();
+    if (start_result.is_err()) {
+        pacs::samples::print_error("Failed to start server: " +
+                                   start_result.error().message);
+        return 1;
+    }
+
+    // Print server running banner
+    pacs::samples::print_box({
+        "Echo Server Running",
+        "",
+        "Test with DCMTK:",
+        "  echoscu -aec ECHO_SCP localhost " + std::to_string(config.port),
+        "",
+        "Test multiple connections:",
+        "  for i in {1..5}; do echoscu -aec ECHO_SCP localhost " +
+            std::to_string(config.port) + "; done",
+        "",
+        "Press Ctrl+C to stop"
+    });
+
+    // Wait for shutdown signal
+    // This blocks until Ctrl+C is pressed or server->stop() is called
+    sig_handler.wait();
+
+    // ===========================================================================
+    // Part 5: Statistics and Cleanup
+    // ===========================================================================
+    // After shutdown, we can retrieve server statistics for monitoring.
+    // The server automatically cleans up all resources.
+
+    pacs::samples::print_section("Final Statistics");
+
+    auto stats = server->get_statistics();
+
+    pacs::samples::print_table("Server Statistics", {
+        {"Total Associations", std::to_string(stats.total_associations)},
+        {"Rejected Associations", std::to_string(stats.rejected_associations)},
+        {"Messages Processed", std::to_string(stats.messages_processed)},
+        {"Bytes Received", std::to_string(stats.bytes_received)},
+        {"Bytes Sent", std::to_string(stats.bytes_sent)},
+        {"Uptime", std::to_string(stats.uptime().count()) + " seconds"}
+    });
+
+    // Summary
+    pacs::samples::print_box({
+        "Congratulations! You have learned:",
+        "",
+        "1. Server Configuration - AE Title, port, timeouts",
+        "2. Service Registration - SCP services for DIMSE handling",
+        "3. Association Lifecycle - Connect, exchange, release",
+        "4. Event Callbacks - Monitoring and logging",
+        "5. Graceful Shutdown - Clean resource cleanup",
+        "",
+        "Next step: Level 3 - Storage Server (C-STORE SCP)"
+    });
+
+    pacs::samples::print_success("Echo Server terminated successfully.");
+
+    return 0;
+}

--- a/samples/CMakeLists.txt
+++ b/samples/CMakeLists.txt
@@ -11,7 +11,7 @@ add_subdirectory(common)
 
 # Sample applications
 add_subdirectory(01_hello_dicom)       # Level 1: Basic DICOM file operations
-# add_subdirectory(02_echo_server)     # Level 2: C-ECHO SCP/SCU
+add_subdirectory(02_echo_server)       # Level 2: C-ECHO SCP/SCU
 # add_subdirectory(03_storage_server)  # Level 3: C-STORE SCP/SCU
 # add_subdirectory(04_mini_pacs)       # Level 4: Full PACS with query/retrieve
 # add_subdirectory(05_production_pacs) # Level 5: Production-ready features

